cmd/ocsp-updater/main.go

```diff
@@ -107,97 +203,117 @@ func (updater *OCSPUpdater) processResponse(tx *gorp.Transaction, serial string)
+				tx.Rollback()
+				return
+			}
+		}
+		err = tx.Commit()
+		if err != nil {
+			updater.log.AuditErr(err)
+			return
+		}
+	}()
+
+	return
+}
+
+// oldOCSPResponsesTick looks for certificates with stale OCSP responses and
+// generates/stores new ones
+func (updater *OCSPUpdater) oldOCSPResponsesTick(batchSize int, now, _ time.Time) {
+	updater.rpcMu.RLock()
+
+	certs, err := updater.findStaleOCSPResponses(now.Add(-updater.ocspMinTimeToExpiry), batchSize)
+	if err != nil && err != errNoStaleResponses {
 		updater.stats.Inc("OCSP.Updates.Failed", 1, 1.0)
-		tx.Rollback()
-		return err
+		updater.log.AuditErr(err)
+		updater.rpcMu.RUnlock()
+		return
 	}
 
-	updater.log.Info(fmt.Sprintf("OCSP %s: OK", serial))
-	updater.stats.Inc("OCSP.Updates.Processed", 1, 1.0)
-	updater.stats.TimingDuration("OCSP.Updates.UpdateLatency", time.Since(innerStart), 1.0)
-	return nil
+	wg := new(sync.WaitGroup)
+	updater.generateOCSPResponses(wg, certs)
+	wg.Wait()
+	updater.rpcMu.RUnlock()
 }
 
-// findStaleResponses opens a transaction and processes up to responseLimit
-// responses in a single batch. The responseLimit should be relatively small,
-// so as to limit the chance of the transaction failing due to concurrent
-// updates.
-func (updater *OCSPUpdater) findStaleResponses(oldestLastUpdatedTime time.Time, responseLimit int) error {
-	var certificateStatus []core.CertificateStatus
-	_, err := updater.dbMap.Select(&certificateStatus,
-		`SELECT cs.* FROM certificateStatus AS cs JOIN certificates AS cert ON cs.serial = cert.serial
-		 WHERE cs.ocspLastUpdated < ? AND cert.expires > now()
-		 ORDER BY cs.ocspLastUpdated ASC
-		 LIMIT ?`, oldestLastUpdatedTime, responseLimit)
-
-	if err == sql.ErrNoRows {
-		updater.log.Info("All up to date. No OCSP responses needed.")
-	} else if err != nil {
-		updater.log.Err(fmt.Sprintf("Error loading certificate status: %s", err))
-	} else {
-		updater.log.Info(fmt.Sprintf("Processing OCSP Responses...\n"))
-		outerStart := time.Now()
-
-		for i, status := range certificateStatus {
-			updater.log.Debug(fmt.Sprintf("OCSP %s: (%d/%d)", status.Serial, i, responseLimit))
-
-			err = updater.updateOneSerial(status.Serial)
-			// Abort if we recieve a fatal error
-			if _, ok := err.(FatalError); ok {
-				return err
-			}
-		}
+type looper struct {
+	clk       clock.Clock
+	stats     statsd.Statter
+	batchSize int
+	tickDur   time.Duration
+	tickFunc  func(int, time.Time, time.Time)
+	name      string
+}
 
-		updater.stats.TimingDuration("OCSP.Updates.BatchLatency", time.Since(outerStart), 1.0)
-		updater.stats.Inc("OCSP.Updates.BatchesProcessed", 1, 1.0)
+func (l *looper) loop() {
+	prev := l.clk.Now().Add(-l.tickDur)
+	for {
+		now := l.clk.Now()
+		l.tickFunc(l.batchSize, now, prev)
+		l.stats.TimingDuration(fmt.Sprintf("OCSP.%s.TickDuration", l.name), time.Since(now), 1.0)
+		// If the tick didn't take as long as expected sleep for the difference
+		if diff := l.clk.Now().Sub(now.Add(l.tickDur)); diff > 0 {
```

Comment:
I think this would be better as:
```
expectedTickEnd := tickStart.Add(l.tickDur)
longTickEnd := tickStart.Add(3 * l.tickDur)
actualTickEnd := l.clk.Now()
if actualTickEnd.Before(expectedTickEnd) {
} else if actualTickEnd.After(longTickEnd) {
```

Also, I think the stat names are a little odd. What you have labeled as `ShortTicks` is actually the expected case, right? All work should be done before it's time for the next tick? I would probably just break these into `Ticks` (counting all types) and `LongTicks` (taking longer than `tickDur`).